added support for delayed phonetic character descriptions

davidacm opened the pull request:
### Link to issue number:
Closes #13509

### Summary of the issue:
The main use case is navigating character by character and a user encounters a character that they do not understand. The delayed phonetic character description is something like “alpha”/“beta”/“romeo”/“igloo”.

**What makes understanding a character difficult?**
This may be because a synthesizer pronounces a character unusually or indistinguishable from another character.
A user may also be hearing impaired or have other trouble processing the character. In this case, the user waits for the delayed phonetic description to occur.

Other ways to report the character description:
- script_review_currentCharacter
- script_review_currentWord
- script_review_currentLine
- script_reportCurrentLine

**What makes determining the delay difficult?**
The main determinate of the delay is the speed that a user interacts with the device, e.g. speed of navigating by character.
NVDA currently does not have a way to configure interaction rate, which is a general problem described in #13915.

Meeting auditory processing needs is generally achieved by adjusting the synthesizer rate.
Unfortunately, synthesizer rate varies by synthesizer, and so calculating a delay based on rate is challenging.

While a general solution for the interaction rate is blocked by #13915, using a sensible default will handle most users requirements. This delay can be configured easily by add-ons.

### Description of user facing changes
Adds an option, which when enabled, causes character descriptions to be announced after a fixed delay of 1 second.
This delay value has been tested as a default in the “Enhanced phonetic reading” Add-on.

### Description of development
This code modifies getTextInfoSpeech to emit a BreakCommand followed by several commands to speak a character description. This only happens if the unit of resolution is a single character, and the reason for querying for textInfo speech is caret. This causes NVDA to emit character descriptions in the following conditions:

* when moving by character in an edit field.
* When moving by character in review cursors.
* When moving by character in browse mode.
* When reporting single characters at review cursor.
* Any other time that caret movement is being reported for single characters.

This solution also reads the character description with capitalization settings. 

Using the BreakCommand means that the overall approach is simpler as opposed to using a timer, and handled by the synthesizer rather than NVDA.

### Testing strategy:
This feature was taken from an add-on called "Enhanced Phonetic Reading". The UX has been tested by many users in form of add-on by years. The default delay is 1 second, and can be configured between 100-5000ms. 
The internal code has largely been rewritten in this PR.

System tests have been created to test that delayed descriptions are spoken.

Manual testing of reading characters in notepad has been performed using eSpeak, OneCore, SAPI5 and SAPI4.
This involved testing different delay rates and cancelling the description.

### Known issues with pull request:
None

### Change log entries:
New features

- Added ability to read character descriptions when the user is moving by characters, delayed by 1 second.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

Commit:
Added support for typed characters to NVDA.

Commit:
Update gui and user documentation.

Commit:
I don't even know, okay!

Commit:
Merge branch 'master' of https://github.com/nvaccess/nvda into typed

Commit:
Remove random trailing whitespace removals.

davidacm committed:
added support for delayed phonetic descriptions when moving the cursor by characters.

davidacm committed:
fix now NVDA won't try to get character descriptions if info.text is empty.

derekriemer committed:
Update source/gui/settingsDialogs.py

Co-authored-by: Ozancan Karataş <ozancankaratas96@outlook.com>

derekriemer committed:
Update source/gui/settingsDialogs.py

Co-authored-by: Ozancan Karataş <ozancankaratas96@outlook.com>

derekriemer committed:
Update source/config/configSpec.py

Co-authored-by: Ozancan Karataş <ozancankaratas96@outlook.com>

cary-rowen commented:
I am interested in this feature, But the current CI build is dead and I can't test it.
can you provide a new one

davidacm committed:
fix linting errors from flake8

davidacm commented:
Hi @cary-rowen , the links expire after one month, try with this new one that was generated by the new commit. Please take in account that your antivirus can try to delete this file because the installer is unsigned.
[]NVDA build with delaied descriptions(https://ci.appveyor.com/api/buildjobs/bkjt0rnrc065vvno/artifacts/output%2Fnvda_snapshot_pr13550-25331%2C7afa5930.exe)

cary-rowen commented:
Hi @davidacm
I tested it,
For the "Time for delayed phonetic descriptions" control, can it be displayed or not based on the selected state of "Delayed phonetic descriptions for characters on cursor movement"?

davidacm committed:
added feature to hide and show the field delayedPhoneticDescriptionsTimeoutMsEdit if delayedPhoneticDescriptionsCheckBox is toggled.

davidacm commented:
Hi @cary-rowen.
Thanks for the suggestion. I implemented it now, please try to see if it works for you!

Thanks!

LittleStar-VIP commented:
Hello,
I have downloaded https://ci.appveyor.com/api/buildjobs/bkjt0rnrc065vvno/artifacts/output%2Fnvda_snapshot_pr13550-25331%2C7afa5930.exe and tested it in Notepad++, Brave web browser, Powershell, and everything is as expected.
Once again, may I thank you again for this wonderful addon and feature.

cary-rowen commented:
Hi, @davidacm 
Can you keep improving this pr, I'm looking forward to him appearing in NVDA core.

davidacm commented:
Hi @cary-rowen , I don't know what more improvements I can do to this pr. I have some ideas related to this feature, but those ideas should be implemented in another pr because I don't want to mix features.

I'm open to suggestions in code implementation and improvements, and ideas from users, but at this time, I did receive suggestions about that.

cary-rowen commented:
Hi @davidacm 

AppVeyor CI seems to have a check failure, you might be able to take a look,

---

What is NV Access's attitude towards this, can you take a look at this PR?
@seanbudd 
@feerrenrut

davidacm commented:
its a System failure, but I don't understand why, if the first time the system test was passed, just linting errors were detected but they were corrected in the last commit.

seanbudd commented:
Sorry @davidacm - this slipped of my radar when dealing with the latest 2022.1 release. I will look into reviewing this ASAP

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -1252,6 +1252,19 @@ This option should generally be enabled.
 If you are having problems with the pronunciation of individual characters, try disabling this option.
 
+==== Delayed phonetic descriptions for characters on cursor movement ====[delayedPhoneticDescriptions]
+This setting is a checkbox that, when checked, tells NVDA to say the phonetic description when you move by characters, after an user defined pause. Default delay setting is 1000 ms (1 sec).
```
Adding newlines between sentences helps translators
```suggestion
When this setting is checked, NVDA will say the phonetic description when you move by characters, after an user defined pause.
Default delay setting is 1000 ms (1 sec).
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -1252,6 +1252,19 @@ This option should generally be enabled.
+
+This can be useful if your current speech synthesizer pronounces some characters in a similar way. Also, can be useful if you have some degree of deafness. In this case, could be a good idea to set NVDA to use small delays.
+
+==== Time for delayed phonetic descriptions (ms) ====[delayedPhoneticDescriptionsTimeoutMs]
+This edit field allows you to type the pause time that NVDA will take to say the phonetic description when you move by characters.
+This setting is specified in milliseconds (ms).
```
```suggestion
==== Time for delayed phonetic descriptions (ms) ====[delayedPhoneticDescriptionsTimeoutMs]
Specify the delay time in milliseconds (ms) to say the phonetic description when you move by characters.
Used when delayed phonetic descriptions is enabled.
```

seanbudd commented on source/speech/speech.py:
```diff
@@ -2595,3 +2610,46 @@ def clearTypedWordBuffer() -> None:
+
+def cancelPhoneticDescriptionTimer() -> None:
+	"""
+	this stops the timer used for delaied phonetic descriptions.
+	This should be called when a new sentence is send or the user stops the synth.
+	E.G, by pressing control key.
```
```suggestion
	Stops the timer used for delayed phonetic descriptions.
	This should be called when a new sentence is sent or the user cancels the speech.
	e.g, by pressing control key.
```

seanbudd commented on source/speech/speech.py:
```diff
@@ -810,6 +812,8 @@ def speak(  # noqa: C901
 
 	if not speechSequence:  # Pointless - nothing to speak
 		return
+	# cancel any delaied phonetic description if exists.
```
```suggestion
```

seanbudd commented on source/speech/speech.py:
```diff
@@ -2595,3 +2610,46 @@ def clearTypedWordBuffer() -> None:
+
+
+def speakDelayedDescription(info: _FakeTextInfo):
+	"""
+	this function is used to announce the delayed descriptions.
+	We can't call spellTextInfo directly because we need to check if the description is available first.
+	"""
```
```suggestion
```

seanbudd commented on source/speech/speech.py:
```diff
@@ -2595,3 +2610,46 @@ def clearTypedWordBuffer() -> None:
 	_curWordChars.clear()
+
+
+def cancelPhoneticDescriptionTimer() -> None:
```
All of this new code should be added to a new file `source/speech/phoneticDescriptions.py`, and imported into here.

seanbudd reviewed (commented):
Sorry for the late review, I've added some suggestions. I think the approach generally looks good but the documentation and design needs some work.

seanbudd commented:
It would be great if there were system tests for this behaviour: in `tests\system\robot\symbolPronunciationTests.py`

Some things to test could include:
- different delay periods (min, max, default) are roughly accurate
- cancelling works with other speech
- cancelling works with control

cary-rowen commented:
@davidacm 
just a reminder.

Thanks

davidacm committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm committed:
Update source/speech/speech.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

davidacm commented:
Hi @seanbudd.
Thanks for the suggestions :) I'm working on this, currently I have a question:
I'd like to know an example of synthesizer where this feature would not work properly. Maybe I can look for a solution on this?
This feature is like read character descriptions, but with a timer, the timer is internal, does not rely on a synth timer. So, i though if the common read character description works properly, this feature should work as well.

I agree with all suggested changes.

seanbudd commented:
Thanks @davidacm 

>I'd like to know an example of synthesizer where this feature would not work properly. Maybe I can look for a solution on this?
This feature is like read character descriptions, but with a timer, the timer is internal, does not rely on a synth timer. So, i though if the common read character description works properly, this feature should work as well.

I'm not sure what this question is asking - I figure it is in response to the request for system tests.

I'm not aware of any examples of a synthesizers where this feature would not work.
The system tests use a fake synthesizer and do not test against real synthesizers.
It's not up to NVDA to test synthesizers - it is up to the synthesizers to test themselves.
However, we do test that the information we send to synthesizers (i.e. the integration) is correct.

System tests are helpful for discovering unknown issues. If there was a known issue with your code, you would be asked to fix the issue rather than make system tests.
Adding systems will clearly document how your feature works, as well as ensure future work on the feature doesn't break it.

davidacm commented:
Hi @seanbudd.
I was trying to move the functionality to another file as you suggested. But since this file would need to be imported in speech.py, and this funcionality requires two methods from speech.py... this can generate circular imports.
I can import those functions inside a function to skip this limitation, but I don't know if is a good idea.
Functions are the followin:

* getCurrentLanguage()
* spellTextInfo(info, useCharacterDescriptions=True)

What do you recommend?

Thanks in advance!

seanbudd commented:
>I can import those functions inside a function to skip this limitation, but I don't know if is a good idea.

That should be fine. In future we can pull more functions out of `speech/speech.py` to reduce the occurrence of circular dependencies.

davidacm committed:
added a new file to manage the delayed character descriptions.
changed all references of word "phonetic" to character when referring to delayed character descriptions to be consistent with NVDA documentation. It was called delayed phonetic descriptions before this commit.
changed the name of the function cancelDelayedCharacterDescriptionIfPending to "cancelDelayedCharacterDescription" to avoid very long names. The behavior of the function is now further documented to clarify that it will validate if a delayed character description is pending, so this function can be called at any time without problems.

davidacm committed:
added tests for delayed character descriptions.
tests are done with min delay, max delay, default delay.
also, a test is interrupted by the control key, and by reading another sentence (the title of the application)

davidacm commented:
Hi, I'm having problems to execute the robot tests on my local machine due to the environment, and not enough time to solve them soon. Also my inexperience doing robot tests for NVDA...
Can anyone execute the robot tests present in the latest commit for the symbolPronunciationTests?
I will try to fix my environment in some days, after some personal activities that I have to do first.

Or if there was a way to make them without relying on visual studio, the cause of my local issues. I'm open to suggestions.

davidacm committed:
small fix when getting test for delayed desccriptions.
Fix a mistake when using Move enum class in symbolPronunciationTests

davidacm committed:
tests for delayed descriptions have been implemented. Some issues may occur due to the latency in communication between robot framework and NVDA. if you are experimenting issues, see the NVDA log to ensure that the times between a read character and the description is OK. You can increase the threshold tolerance if that occur.

davidacm committed:
code style corrections.

davidacm commented:
Hi @seanbudd ,
I finally implemented system robot tests for delayed descriptions.
But some drawbacks occurred due to te way that the robot framework is working.

Is hard to measure time accurately for some reasons:
E.G. the events are checking by observing a function by intervals, the latency between the client and the server, etc.
Tests of stop speaking before the description is read, and send another command to NVDA before the description, are working without problems.
Timing measures may vary between runs even by 100ms, but when I check the internal logs, the times are OK with the specified settings.
I reimplemented some functions to mitigate the inconvenience.

P.S. Sorry for the delay, I was with covid and other things these weeks.

davidacm commented:
BTW, another inconvenience is that the character descriptions used may change according to the user system language.

davidacm committed:
added an original function that was accidentally removed.

seanbudd committed:
Merge branch 'master' into delayedDesc

seanbudd committed:
review actions

seanbudd commented:
Hi @davidacm -

I've pushed a large number of changes.
Would you be able to review these and let me know if there are any issues?
I will annotate the diff to provide some context.

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -1618,8 +1618,10 @@ def makeSettings(self, settingsSizer):
-		# voice settings panel.
+		self._appendDelayedCharacterDescriptions(settingsSizerHelper)
+
+	def _appendDelayedCharacterDescriptions(self, settingsSizerHelper: guiHelper.BoxSizerHelper) -> None:
```
I've moved this code into it's own function for modularity

seanbudd commented on source/gui/settingsDialogs.py:
```diff
@@ -1640,27 +1642,28 @@ def makeSettings(self, settingsSizer):
+			self.delayedCharacterDescriptionsTimeoutMsEdit.Disable()
 
-	def onToggleDelayDescriptions(self, evt):
+	def onToggleDelayDescriptions(self, evt: wx.CommandEvent) -> None:
```
This should enable/disable rather than show/hide.
Show/hide causes visual formatting issues and can be confusing. Enable/disable makes it clear what selecting the checkbox does.

seanbudd commented on source/speech/delayedCharacterDescriptions.py:
```diff
@@ -42,29 +39,27 @@ def cancelDelayedCharacterDescription() -> None:
-	# We can't call spellTextInfo directly because we need to check if the description is available first.
 	if description:
-		spellTextInfo(info, useCharacterDescriptions=True)
+		_spellTextWithFields(info, useCharacterDescriptions=True)
```
if `spellTextInfo` is ever updated to use other aspects of a `TextInfo`, calling it with a `_DelayedCharacterDescriptionTextInfo` may cause an error. As such, I have split out `_spellTextWithFields` from `spellTextInfo`. `_spellTextWithFields` supports the shared Protocol `_SupportsGetTextWithFields` used by both `TextInfo` and `_DelayedCharacterDescriptionTextInfo`.

seanbudd commented on source/speech/delayedCharacterDescriptions.py:
```diff
@@ -42,29 +39,27 @@ def cancelDelayedCharacterDescription() -> None:
-		for k in info.fields:
-			if isinstance(k, textInfos.FieldCommand) and k.command == "formatChange":
-				curLang = k.field.get('language', curLang)
+		for command in info.getTextWithFields():
```
variable names should be descriptive

seanbudd commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -313,25 +309,11 @@ def test_moveByChar():
 }
 
 
-def _wait_for_specific_speech(speech: str, maxWaitSeconds: int) -> bool:
```
This has been removed in favour of improving the API of `spy.wait_for_specific_speech`

seanbudd commented on tests/system/robot/symbolPronunciationTests.py:
```diff
@@ -353,111 +335,109 @@ def _getCharAfterKey(key, maxWait=0.1) -> str:
 
-def _testDelayedDescription(key: Move, delay: int = 1000, repeatCount: int = 1, threshold: int = 70,) -> bool:
-	""" perform delayed character descriptions tests with with the specified parameters:
+def _testDelayedDescription(key: Move, expectedDelaySeconds: float, thresholdSeconds: float = 1) -> None:
```
to simplify these tests:

- I have removed the repeat count, performing a test successfully once should suffice.
- Upped the threshold. Devices will vary in performance. 1s max delay is acceptable (this allows a user to set the delay to 50ms and experience at most a delay of ~1s). As mentioned in other comments, the framework and measuring the delay makes up most of that threshold.

davidacm replied:
Hi @seanbudd, all changes are perfect for me. So, I don't have more things to add on this pr.
Thanks for the suggestions, I like this new way in testing robot, thanks for the improved wait_for_specific_speech function.

Let me know if you want me to do something else in this pr.

Regards.

seanbudd committed:
Fix lint

seanbudd committed:
Use BreakCommand instead of timers

seanbudd committed:
fix lint

seanbudd committed:
fix up comment

seanbudd committed:
add typing to charProcessing

seanbudd commented:
@davidacm - we've decided to go with method used in #13543 to speak the character descriptions, using the `BreakCommand` instead of timers.

The issue with eSpeak and the `BreakCommand` has been fixed with a workaround while we wait for https://github.com/espeak-ng/espeak-ng/issues/1232 to be fixed.
Using the `BreakCommand` means that the overall approach is simpler, and handled by the synthesizer rather than NVDA.

The only issue with the new approach is that synthesizers must support cancelling speech and the `BreakCommand`.
`SpeechSpySynthDriver` used in system tests doesn't support either of these, and thus the system tests will fail until this is implemented.

seanbudd committed:
remove default delay from userGuide

it should be independent

davidacm commented:
Hi @seanbudd. I don't agree with #13543 method, because Espeak is not the unique synthesizer and users may be using a tts that does not support break command. Consider not english only, but there are many languages and synthesizers in the world.
Although I agree with the decision.

Regards.

seanbudd commented:
@davidacm - Synth drivers should implement a `BreakCommand` if they want to implement this feature.
If a synth driver does not implement the `BreakCommand` they will miss out on all other instances of NVDA using the command, this is not a unique problem to this PR.
This feature works with OneCore, eSpeak and SAPI5.

davidacm commented:
Hi @seanbudd,
This is not intended to be a discussion, but a concern of mine and of some people who expressed their concern to me by mail a couple of months ago. Please read arguments carefully, not just thinking about whether a code is more or less complex.

### Cons of using break commands:

1. it depends if the synth library supports it, or if the developer implemented it on the NVDA synth driver. It's no way to control it, we can't force developers to implement it.
2. Also, we should not force the users to find other TTS just because the synth they are using, don't support break command. We can't force users to use or avoid the use of a synth. As a developer, I like to think in all possible cases, even if I don't know of one particular case of this. That's why we have many inaccessible applications, because those developers did not think about blind people and I always try not to make the same exclusion of users.
3. We don't know all NVDA's communities. Just to mention some examples, the Chinese, Japanese or Arabic communities are not very well known to me and I don-t know what tts synths are they using, or if those synths supports some features or not.
4. Lastly, some tts synths are not accurate in the break pauses, depending of the rate adjustment of the synthesizer. We can rely on the modern synths, but many users are using old voices yet.

### Cons of using timers:

A little more code complexity. is the only drawback I can think of.

We don't need to go far to find examples of synthesizers that don't support break command feature.
nvSpeechPlayer, picoTTS and even sapi4 driver, don't support it. Just to mention some examples of very popular synths that will be excluded of this feature.
Sapi 4 is a good example, many people are  still using sapi 4 to use old but good voices.

Finally, I agree that is less complex to use break commands, but a timer is not too complex I would say. And it's very accurate.
Break command approach won't affect to me, or majority of english / spanish / portuguese  people etc, but may leave out many users who may not be as well known in the community. While timer approach will work for every synthesizer in the world, without changing anything in the code of add-ons. Just with a small extra complexity added. I don't see any cons other than that.

I know that there are another features of NVDA that are using break commands, and then won't work with those synths, but this feature could be one less.

I would say that these are my concerns if break commands are used, but of course I respect the decision you have made.

P.S. English is not my native language. Sorry if had mistakes or if some of my words sound harsh or rude. It would never be my intention.

You don't need to answer or justify my arguments or this comment, I just wanted to express what I think in this area.

Regards.

cary-rowen commented:
@seanbudd 
I totally agree with @davidacm, I am a Chinese user and there are many people in the Chinese community who are still using the SAPI4 speech synthesizer.

seanbudd commented:
@cary-rowen are you sure SAPI4 does not support the break command? I'm unable to test it

davidacm commented:
I was unable to test sapi 4 on my system, it doesn't want to work for me. But I can't see references to break command in the sapi4 driver.

but at best with sapi 4 supporting this, there will still be many synthesizers that don't support it. And we shouldn't leave them out, since the solution to make this feature generalized for all synths is as simple as using timers instead of breaks.
There could be some drivers out there of importance to users that probably won't be updated to support break commands.

And the other issue, synthesizers that change the pause time according to the rate setting.

XLTechie commented:
@seanbudd what is the advantage in guaranteeing a differential user experience 
across different synths? Is the timer approach really that much more 
problematic?

The break method is nice from a code purity and "way it should be done" 
prospective, but shouldn't the over all user experience of the product be of 
paramount importance, especially when the difference is of low operational 
impact?

I am imagining users on mailing lists asking why this feature isn't working for 
them, and the back and forth debug process until somebody realizes their using 
an unsupported synth.

What's worse, it means that the add-on which originally implemented this 
feature, must still be maintained, distributed, recommended, and used, but now 
only for certain synths.

In my mind, there is no point in bringing a popular add-on feature across to 
core, if it is not done in a way that eliminates the need for the add-on by the 
incalculable majority of users.

(Certainly incalculable from here, since only NV Access has access to the pushed 
stats that might indicate which synths are in use globally.)

davidacm commented:
Hi again, I finally got sapi 4 working on my computer. And not, break commands are not working on sapi 4, then the delay can't be adjusted.
The character description is announced just after the character is read.
But I found other interesting things:

Break commands are not accurate on sapi 5, one core voices, and even can be a little inaccurate on Espeak.
Do the following tests: read a character with te lowest rate on one core or sapi 5, and then, increase the speed. You can notice that the delay is lower if you increment the voice rate speed, even with small changes.
You can observe it on Espeak too, but you need to do larger speed increments, so that the difference is noticeable to the ear.

P.S. I think that 500ms by default is a little fast for a character description. I hadn't noticed it because I hadn't tried the new commits in depth. With 500ms, can be tedious for the users.

davidacm commented:
correction: You can notice that the delay is faster if you increment the voice rate speed.
This happen with one core, sapi 5, and espeak.
In sapi 4, break command doesn't work at all.

seanbudd commented:
>what is the advantage in guaranteeing a differential user experience across different synths?
I believe a different UX across synthesizers is unavoidable, and inherent to existence of separate synthesizers.
Ideally all Synth Drivers using NVDA support the `BreakCommand`.

>if it is not done in a way that eliminates the need for the add-on by the incalculable majority of users.

I really challenge this assertion. With #13875 all included synthesizers in NVDA support the `BreakCommand`.

The BreakCommand API has been around for a long time. By introducing this new feature, we are choosing to do it a way that is standard for the NVDA core API, which is better for long term support of the feature by both external synthesizers and NVDA core.
If external add-ons are yet to opt-in to this API, they can do so at any time with their own implementation.

seanbudd commented:
>especially when the difference is of low operational impact

I also strongly challenge this assertion, this change has now avoided:
* needing extensive system tests that check for timing and cancellation, to trusting the BreakCommand and synthesizers
* needing a new module
* creating a new design pattern just for timer based callback announcements, when generally in NVDA we just announce more details after an EndUtterance or similar delay
* having to create a fake text info interface and splitting up `spellTextInfo`
* increasing reliance on `wx.CallLater`/`core.CallLater`, which is a really problematic API as it stands

seanbudd committed:
remove eSpeak fixes to different PR

seanbudd committed:
revert default to 1000ms

seanbudd committed:
simplify system tests

seanbudd committed:
split out system test framework change to separate PR

seanbudd committed:
split out typing fixes to separate PR

seanbudd commented:
@davidacm 
>If it's on your hand, can you describe the inconvenience of using wx.CallLater. This can be useful for add-on developers that are considering to use this API.
>I asked for this some months ago, when I replaced python timers to avoid incompatibilities with other add-on, but no one replied to my question.

Could you point towards that question?
It would be better to answer this on a GitHub discussion or issue directly, rather than in the comments of this PR

>the differences in the pause time that are observed in sapi 5 and one core when the speed rate is changed, can be fixed too?

I imagine this is a problem with OneCore / SAPI5 and needs to be reported to Microsoft.
We can investigate this, but generic problems with the BreakCommand is something that should be addressed in separate issues rather than this PR.
I don't think the inaccuracy of this should block this PR. With this PR, the delay time can be configured on a synth by synth basis.
If users are changing speech rates for a synthesizer, they might even prefer the delay to scale with it, though I agree this should be documented if that is the case. Otherwise, you can always change the delay time when you change the speech rate.

Could you explain why an accurate specific delay time is important for the user?
To make this PR simpler, I could imagine just using the fixed rate of 1 second, or short/medium/long pauses as options. I am surprised by the need to configure it to such a specific degree.

seanbudd committed:
add Derek as contributor

XLTechie commented:
@seanbudd wrote:

> Could you explain why an accurate specific delay time is important for the user?
> To make this PR simpler, I could imagine just using the fixed rate of 1 second, or short/medium/long pauses as options. I am surprised by the need to
> configure it to such a specific degree.

Personally I never quite understood why the ms specification from both of the 
rival PRs to fix this was necessary; I always thought the short/medium/long 
model would be more user friendly. I believe I even said so on @derekriemer's 
PR.

For those of us, like me, who are very picky about how our speech works, I can 
see that choosing the exact number of mills that we are comfortable with would 
be valuable, but is that really most users?

I would like it at least somewhat configurable though. For those who use high 
speech rates with ESpeak, etc., having to wait a full second for the description 
to speak would be an annoyance. On the other hand, those who use OneCore at a 
slow rate, may find 300 ms to be far too short and prefer the full 1000.

My preference for this would be short, medium, and long, with the exact ms 
values set by config.conf variables that aren't exposed in the UI, but are 
available to power users as other things are.

seanbudd commented:
@XLTechie - I'd like to highlight that scaling a fixed delay with speech rate is also an option. This would remove the need to specify delay as an option, but might take some time to get the scaling right.

davidacm commented:
The end user's opinion and comfort is what should matter most, rather than whether you or I want to implement this one way or the other as a matter of personal preference.

As user I think that delay time accuracy is important in my opinion, and that was the main reason about create the tests. why?
Because for me, as an user using a high rate speeds, even for me is tedious that the delayed description is short. I prefer the way like voice over, JAWS or talkback works, that the delay doesn't change if I update the rate speed. We are not thinking about how this might affect a person with hearing loss, for example. Some of them use very short times and the change might be more noticeable when updating the speed. I can't speak for them, but I wouldn't go so far as to say that accuracy is not important either.
I change the speed very often, it depends if I am reading, or if I am using headphones or speakers, if the day is rainy, etc. But short delays can be tedious when reading by characters. I mean, that as an user I don't want to be changing the time delay every time I change the speed.
I think that may be a good idea to ask the final users if the delay accuracy is important or not. It's important for me, but I can talk for all the users in the world.

This accuracy can be tedious on sapi 5 and one core voices, where the differences between pause time and the speed changes, are very noticeable.
I doubt that microsoft will correct this soon, at least in sapi 5, where most of the voices are. We can either wait in our sleep, or look for another approach.

About the CallLater: I asked it on the nvda community some months ago, but I didn't get a response, so, when this topic come here, I thought that the answer could be clarified. I don't see why we can discuss about it here, and translate it to the nvda developer list, is not hard to answer if one have the knowledge about CallLater. But I will ask to answer there again, I guess. I'm really interested as a NVDA add-on developer and an user of the wx framework in other projects.

LittleStar-VIP commented:
>  Could you explain why an accurate specific delay time is important for the user?

Before answering this question, I think I could also ask  
Could you explain why an accurate specific pitch level is important for the user?
Could you explain why an accurate specific speaking rate is important for the user?
etc.
So we can also just offer three options to users regarding speaking rate, pitch level etc.

NVDA is a screen reader that support multiple synth and multiple language.
Specific delay may work for specific language or synth but not work well for the other synth or language. If you think that those three options that you will be offering can fit into different situations, then I have nothign to say.

As a user, I do not know much about technical issue, but I would strongly prefer the option to configure specific delay time in ms.
I remember when I was first using the original enhance phonetic addon, I spend some time to adjust the delay time based on the synth that I was using, instead of the default option.

beqabeqa473 commented:
I'll give my 5 cents too.

I think delay should not be bound on speed change and should be configurable.
It shouldn't matter how fast or slow my speech is, delay should be
constant in all cases.
If i need shorter or longer pause, i should have a possibility to configure it.

thanks.

On 7/8/22, LITTLE STAR☆VIP ***@***.***> wrote:
>>  Could you explain why an accurate specific delay time is important for
>> the user?
>
> Before answering this question, I think I could also ask
> Could you explain why an accurate specific pitch level is important for the
> user?
> Could you explain why an accurate specific speaking rate is important for
> the user?
> etc.
> So we can also just offer three options to users regarding speaking rate,
> pitch level etc.
>
> NVDA is a screen reader that support multiple synth and multiple language.
> Specific delay may work for specific language or synth but not work well for
> the other synth or language. If you think that those three options that you
> will be offering can fit into different situations, then I have nothign to
> say.
>
> As a user, I do not know much about technical issue, but I would strongly
> prefer the option to configure specific delay time in ms.
> I remember when I was first using the original enhance phonetic addon, I
> spend some time to adjust the delay time based on the synth that I was
> using, instead of the default option.
>
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/13550#issuecomment-1178622633
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

derekriemer commented on source/speech/speech.py:
```diff
@@ -338,6 +339,40 @@ def _getSpellingSpeechWithoutCharMode(
+		locale: Optional[str] = None,
+) -> Generator[SequenceItemT, None, None]:
+	# This should only be used for single chars.
+	if not len(text) == 1:
```
Does len(foo)==1 work for langs with chars that are not ascii in all cases?

seanbudd replied:
Yes

>The length of a single Unicode character as a Python str will always be 1, no matter how many bytes it occupies.

[Source with example you can test](https://realpython.com/python-encodings-guide/#one-byte-two-bytes-three-bytes-four)

seanbudd commented:
@davidacm 

>As user I think that delay time accuracy is important in my opinion, and that was the main reason about create the tests. why?

The tests were to ensure the new design pattern using the timer system worked properly, and even so it was only to a threshold accuracy of 1 second.

>I think that may be a good idea to ask the final users if the delay accuracy is important or not. It's important for me, but I can talk for all the users in the world.

There's really no way to get a good answer here without users testing the feature on alpha.
If the slight changes in delay time are a big issue, users will raise it.

>About the CallLater: I asked it on the nvda community some months ago, but I didn't get a response, so, when this topic come here, I thought that the answer could be clarified. I don't see why we can discuss about it here, and translate it to the nvda developer list, is not hard to answer if one have the knowledge about CallLater. But I will ask to answer there again, I guess. I'm really interested as a NVDA add-on developer and an user of the wx framework in other projects.

Posting it as an issue or GitHub discussion ensures the answer can be found by others, instead of being hidden in a Google Group discussion or in the comments of a PR. I've created this discussion to capture some notes: https://github.com/nvaccess/nvda/discussions/13886

@LittleStar-VIP 

I think you may have misunderstood the discussion somewhat - with anything in the world, including software, we can only be accurate to a certain threshold. Pitch and rate is handled by the synthesizer, and in this design, so is the delay. It is simply impossible for us to delay something for **exactly** 1 second, so the question becomes what level of accuracy is appropriate. Using either method, the delay setting "1 second" will vary by device performance and the number of tasks NVDA or other processes are trying to handle. For example, if setting the delay to "1 second" roughly translates to 1-1.2 seconds consistently, I really don't see the issue.

LittleStar-VIP commented:
@seanbudd 
I understand that the delay time is not an exact time.

I think the unit of the delay time need not necessary be in microsecond, e.g. it can be just number of second instead of number of microsecond, but I would still prefer to have the option to have a fine adjustment regarding the delay time.

seanbudd commented:
I would like to thank everyone for their feedback on this PR.
The importance of configuring the delay is clear, however we would like to implement a generic solution for this problem.
An outline of this solution can be found in #13915.

In summary, instead of defining a specific delay for phonetic character descriptions, we would like to implement an option to determine how long to wait before considering a user as idle. After no speech or keyboard input after X milliseconds, the delayed phonetic character description would be announced. 

Introducing this generic feature will take longer, and with added risk, so it is not possible to include this in 2022.3.
If we include an option to configure the delay for phonetic character descriptions specifically, it will be difficult to remove when transitioning to the idle time design in a later release (e.g. 2022.4).
As a result, we would like to introduce this feature into 2022.3 with a set delay of 1 second, and introduce the ability to configure the delay in a later release.

Introducing the feature like this also allows add-on authors to easily override how long the delay should be, e.g. with an explicit option, function based on rate, etc.

seanbudd committed:
Use 1 second as fixed delay

seanbudd committed:
Merge branch 'master' into delayedDesc

seanbudd committed:
fix up user guide

Qchristensen reviewed (approved):
User guide change reads well.

seanbudd committed:
fix tests desc

seanbudd committed:
add default to user guide

seanbudd committed:
configure delayed descrpitions for all synthesizers

seanbudd committed:
update changes

seanbudd committed:
fix lint/copy-paste issue

LittleStar-VIP commented:
Thank you every one for making this pr possible, especially @seanbudd  @davidacm

seanbudd committed:
add options to definition

derekriemer commented:
> I would like to thank everyone for their feedback on this PR.
> The importance of configuring the delay is clear, however we would like to
> implement a generic solution for this problem.
> An outline of this solution can be found in #13915
> <https://github.com/nvaccess/nvda/issues/13915>.
>
> In summary, instead of defining a specific delay for phonetic character
> descriptions, we would like to implement an option to determine how long to
> wait before considering a user as idle. After no speech or keyboard input
> after X milliseconds, the delayed phonetic character description would be
> announced.
>
How does this differ from currently, where the delayed description is
canceled after the speech queue is flushed, I.E. a key is pressed or new
speech occurs?

> Introducing this generic feature will take longer, and with added risk, so
> it is not possible to include this in 2022.3.
> If we include an option to configure the delay for phonetic character
> descriptions specifically, it will be difficult to remove when
> transitioning to the idle time design in a later release (e.g. 2022.4).
> As a result, we would like to introduce this feature into 2022.3 with a
> set delay of 1 second, and introduce the ability to configure the delay in
> a later release.
>
> Introducing the feature like this also allows add-on authors to easily
> override how long the delay should be, e.g. with an explicit option,
> function based on rate, etc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/13550#issuecomment-1186729569>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABI2FPP4UZFSY4IZXH2PRCDVUTGS5ANCNFSM5R2YA4KA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Derek Riemer:
Improving the world, one byte at a time. ⠊⠍⠏⠗⠕⠧⠬ ⠮ ⠸⠺⠂ ⠐⠕ ⠃⠽⠞⠑ ⠁⠞ ⠁ ⠐⠞⠲
Software engineer, Drive web

seanbudd commented:
@derekriemer - no difference, except that the delay parameter will be generic, and used to configure other settings as well.

seanbudd committed:
Merge branch 'master' into delayedDesc

davidacm commented:
I don't agree with the delay parameter to be generic if the feature is still relying on the synth break command. Even espeak has imprecision times on rate speed changes, so this can be a issue when changing the synthesizer because with breakCommands, the same value is not equivalent for each synthesizer.
But if the #13915 PR solves the problem of the inaccurate pauses, then I do agree that the parameter should be generic, in fact I find it more useful (at least for me).

I did not see the new code yet, and I didn't see the code of #13915 (although the idea of that pr sounds interesting) at least in the way I understood it.

P.S. Why not reuse the code to read the character description, and the code to obtain the delayed character description?
It's the same thing, the difference is that the first is caused by the user (E.G. by pressing numpad 2 twice) and the second is read after some predefined ms.

I don't know if this has been made, but is just a suggestion.

seanbudd commented:
>P.S. Why not reuse the code to read the character description, and the code to obtain the delayed character description?

This change is simpler, it adds the character description to the same speech sequence, as opposed to starting a new one. The code to read a character description when activating a script is a different scenario - I don't think it makes sense to activate that script as a side effect of speech.

davidacm commented:
I don't mean the script, just the function to get this information. So, both features can call the same function to obtain the information.

P.S. Consider the possibility to use indidual values for the delay for each synthesizer, since most of synths are not accurate in the pause time. I would not like that if i'm using IBMTTS with 1000ms delay and when change to sapi5 or one core, get a different pause time for the delay. Because then, I would need to adjust the delay each time I change the synthesizer.

TechHorseG commented:
I don't know if everything is already locked in, but I would like to register an opinion.

I am very nervous about using a generic setting for the time delay for Delayed descriptions for characters.

To explain. I am used to using a 1 second delay in the Enhanced Phonetic Reading addon, so I would naturally want to initially set this generic idle time delay to 1s.

However, if I then find myself having to wait for 1s delays in other places to get information I used to be able to obtain more quickly, this will be problematic. 

If I have spent years getting the information relatively fast, then I know I am going to find any new delays a drag, and would want to turn them off to retain the quickness I am used to. But then that would be highly problematic for the Delayed descriptions for characters, where I definately do not want an immediate description.

That's it, just wished to register my high feelings of caution on putting the time delay for Delayed descriptions on a generic one-size-fits-all time delay setting, as I feel that this sort of thing has the potential to adversely affect my experience of using NVDA.

On a more positive note. Thanks for introducing this essential feature at all. I haven't updated NVDA in over two years for no other reason than I didn't want to lose the Enhanced Phonetic Reading addon's functionality, when it stopped being compatible.

seanbudd commented:
@TechHorseG - I suggest you read #13915 (and comment there - not here) before making assumptions of what could be affected by a generic parameter.

Additional note: As far I am aware Enhanced Phonetic Reading is compatible with the latest NVDA, and has been regularly updated with each breaking release. Thanks to @davidacm!

davidacm commented:
Hi @TechHorseG.
I'm the developer of the add-on you mentioned.
About enhanced phonetic reading, I will continue with the maintenance until all my concerns about the current implementation in the core of NVDA, are resolved and that may take some years I think.
[Get the latest version of the enhanced phonetic reading here](https://davidacm.github.io/getlatest/gh/davidacm/EnhancedPhoneticReading)
That link always point to the latest release.

Is hard to me to update all my add-ons in the main NVDA's repo, because I still found the process a little tedious, but this add-on will always be available to everyone until it's no longer useful.

You can see my comments about my concerns in this thread. In my opinion timers are the correct way to go, but I know there are some technical limitations and also architecture implementations to be satisfied first.

TechHorseG commented:
Thanks for the enhanced phonetic reading links. I had no idea that the addon was still alive. Good to know!
